File: AspectJCacheExample/src/test/java/com/blogspot/miguelinlas3/aspectj/transaction/TransactionalService.java
package com.blogspot.miguelinlas3.aspectj.transaction;

import com.blogspot.miguelinlas3.aspectj.transaction.marker.Transactional;

public class TransactionalService {

	@Transactional
	public Integer sum(Integer x, Integer y) throws Throwable {
		return x + y;
	}

	public static void main(String[] args) {
		TransactionalService transactionalService = new TransactionalService();

		for (int i = 0; i <= 1; ++i) {
			try {
				transactionalService.sum(i, i);
			} catch (Throwable t ) {
				System.out.println("Exception ocurred on transactional operation." + t.getLocalizedMessage());
			}
		}
	}
}

File: ScalaCacheExample/src/test/scala/com/blogspot/miguelinlas3/scala/transaction/TransactionalService.scala
package com.blogspot.miguelinlas3.scala.transaction

import com.blogspot.miguelinlas3.scala.transaction.manager.{ MockTransactionManager, TransactionManagement }

class TransactionalService(txManager: MockTransactionManager) extends TransactionManagement(txManager) {

  def sum(x: Int, y: Int): Int = {
    transactional[Int]() {
      x + y
    }
  }

  def buggy(x: Int, y: Int): Int = {
    transactional[Int]() {
      val result = x + y
      if (result == 2)
        throw new Exception()
      result
    }
  }
}

object TransactionalService {

  def main(args: Array[String]): Unit = {
    def transactionalService = new TransactionalService(new MockTransactionManager())

    for (index <- 1 to 10) {
      val result = transactionalService.sum(index, index)
      println("Result:" + result)
    }

    for (index <- 1 to 10) {
      try {
        val result = transactionalService.buggy(index, index)
        println("Result:" + result)
      } catch {
        case e => println("An error has ocurred and the transaction has been rolled back")
      }
    }

  }
}